Fix one off in adjust_up and fix out-of-bounds tasks

Jakob37 opened the pull request:
Close  #88

When starting to test d4tools in production, me and @northwestwitch got crashes when calculating many overlapping ranges (i.e. transcripts for a gene).

The crashes do not appear for all genes, but for several, and I could reliably crash separate full datasets using the set of ranges.

Here is an illustration of one of the set of overlapping ranges:

![out](https://github.com/user-attachments/assets/af02f3ce-53e2-42c7-a5d7-75f33aa40db8)

The ranges:

```
1	686509	763194
1	686606	764595
1	686639	764595
1	686659	762714
1	686674	764595
1	686675	764595
1	686682	764595
1	686689	763120
1	686731	762597
1	706399	764595
1	718904	737202
1	747700	748140
```

It seems to be related to the `active_heap`, which is keeping track of parallel ranges. From here, the end ranges are not always retrieved in the right order.

Digging into the `adjust_up` function, called after adding new ranges to `active_heap`. Previously the code checked for parent with a simple division by 2. This will break down for instance for index 2, where the parent will be seen as in index 1. The correct would to be `(idx - 1) / 2` 

See https://en.wikipedia.org/wiki/Binary_heap

>Let n be the number of elements in the heap and i be an arbitrary valid index of the array storing the heap. If the tree root is at index 0, with valid indices 0 through n − 1, then each element a at index i has
>    children at indices 2i + 1 and 2i + 2
    its parent at index [floor](https://en.wikipedia.org/wiki/Floor_function)((i − 1) / 2).

I wanted to do a test case for this, but I have so far not managed to get it running on a small test-dataset. I don't think we need it, but would be good to cross-check this beyond my computer.

As a first step, would you be able to test if this resolves your issues @northwestwitch?

northwestwitch commented:
> As a first step, would you be able to test if this resolves your issues @northwestwitch?

Thanks @Jakob37! Tested on the intervals of my issue:

```
18	29704840	29867058
18	29843485	30050447
18	29847463	30050395
18	29867839	29911633
```

Unfortunately I still get the same error:

```
thread '<unnamed>' panicked at /Users/chiararasi/Documents/work/GITs/d4-format/d4/src/d4file/track.rs:215:21:
attempt to subtract with overflow
```

The file I'm using in the test is the following -> https://d4-format-testing.s3.us-west-1.amazonaws.com/hg002.d4

Jakob37 commented:
> > As a first step, would you be able to test if this resolves your issues @northwestwitch?
> 
> Thanks @Jakob37! Tested on the intervals of my issue:
> 
> ```
> 18	29704840	29867058
> 18	29843485	30050447
> 18	29847463	30050395
> 18	29867839	29911633
> ```
> 
> Unfortunately I still get the same error:
> 
> ```
> thread '<unnamed>' panicked at /Users/chiararasi/Documents/work/GITs/d4-format/d4/src/d4file/track.rs:215:21:
> attempt to subtract with overflow
> ```
> 
> The file I'm using in the test is the following -> https://d4-format-testing.s3.us-west-1.amazonaws.com/hg002.d4

OK! I'll give it a test as well.

Jakob37 commented:
OK, I get the same error for these ranges on my `hg002.d4` file. The debugging continues 🐞

Jakob37 commented:
Hmm. I think there is an additional (separate) layer here, as I think these regions cross the "chunk region" (i.e. at 30000000):

```
# Calculating coverage for:
segment2: 29704840 - 29843485 (1)
segment2: 29843485 - 29847463 (2)
segment: 29847463 - 29867058 (3)
segment2: 29867058 - 29867839 (2)
segment: 29867839 - 29911633 (3)
segment: 29911633 - 30000000 (2)
segment: 30000000 - 29911633 (1)
```

The actual regions:

```
$ cat chiara_regions.tsv
18      29704840        29867058
18      29843485        30050447
18      29847463        30050395
18      29867839        29911633
```

I'll dig around.

Jakob37 commented:
Found it, the region

```
18      29867839        29911633
```

Slips into the subsequent chunk due to the prior ones going there:

```
18      29843485        30050447
```

Just skipping these invalid ranges, it worked for me:

```
                if fpl > r {
                    overlapping_idx += 1;
                    continue;
                }
```

I'll update the PR in a moment so you can give it a try

Jakob37 commented:
Updated! Do you want to give it another go @northwestwitch ?

northwestwitch commented on d4/src/task/context.rs:
```diff
@@ -96,6 +96,14 @@ where
                 }
+                // As ranges are sorted on their left position, the idx approach
+                // above can let through ranges with lower left positions which still
+                // does not reach into the next segment
```
```suggestion
                // do not reach into the next segment
```

northwestwitch commented:
It works fine now! 💯 

```
     Running `target/debug/d4tools stat -s perc_cov=10,15,20,50,100 --region /Users/chiararasi/Documents/work/d4_data/intervals_garem1.bed /Users/chiararasi/Documents/work/d4_data/hg002.d4`

18	29704840	29867058	0.999	0.960	0.797	3.082e-4	0.000e0
18	29843485	30050447	0.999	0.973	0.832	2.271e-4	0.000e0
18	29847463	30050395	0.999	0.972	0.830	2.316e-4	0.000e0
18	29867839	29911633	1.000	0.972	0.833	0.000e0	0.000e0
```

Great fix! I'll close my PR. Let's go for this one!

northwestwitch commented:
@brentp when you have some time could you please review this one? This fix is needed in order to use the new perc_cov stat function. Thanks! 🙏🏻

brentp reviewed (changes requested):
this seems sensible, but can you add a test-case? it can be command-line or from rust.
I'll be a bit of a stickler about tests since we're all in an unfamiliar code-base (though you seem to be getting quite a handle on it).

Jakob37 commented:
> this seems sensible, but can you add a test-case? it can be command-line or from rust. I'll be a bit of a stickler about tests since we're all in an unfamiliar code-base (though you seem to be getting quite a handle on it).

OK, makes sense. Might have to add a larger test file. Will look into it.

Jakob37 commented:
I have added tests for the two errors cases. 

I went for a unit test for the `adjust_up` issue in `tracks.rs`, feeding the same set of ranges that I previously crashed on.

Before

```
~/src/d4-format/d4$ cargo test
...
running 2 tests
test ptab::bit_array::test::test_split_handle_send ... ok
test d4file::track::tests::fail_adjust_up ... FAILED

failures:

---- d4file::track::tests::fail_adjust_up stdout ----
thread 'd4file::track::tests::fail_adjust_up' panicked at d4/src/d4file/track.rs:592:13:
assertion failed: end > begin
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace


failures:
    d4file::track::tests::fail_adjust_up

test result: FAILED. 1 passed; 1 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s

error: test failed, to rerun pass `--lib`
```

After

```
~/src/d4-format/d4$ cargo test
...
running 2 tests
test ptab::bit_array::test::test_split_handle_send ... ok
test d4file::track::tests::fail_adjust_up ... ok

test result: ok. 2 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s

   Doc-tests d4

running 0 tests

test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s
```

For the cross-chunk boundary I added a new d4file with full range chromosomes but mostly without coverage (with coverage  takes some MBs). 

Would have been nice to parameterize the hard-coded chunk size, but that is an adventure for another day.

Before fix

<img width="351" alt="tests_before" src="https://github.com/user-attachments/assets/d2550fbc-9f21-4281-b4fa-ceb2fe700ab8">

After fix

<img width="291" alt="tests_after" src="https://github.com/user-attachments/assets/5fa8a909-77fd-4d66-b74a-1da58056f19f">

Good that you prompted for more tests. I am a bit out of my depth for the unit testing, so if my Googling+ChatGPTing ended in any weird code, just let me know and I'll update.

Jakob37 commented on d4/src/task/perc_cov.rs:
```diff
@@ -11,7 +11,7 @@ pub struct PercentCovPart {
     type ParentType = PercentCov;
     type ResultType = Vec<u32>;
-    fn new(left: u32, right: u32, parent: &Self::ParentType) -> Self {
+    fn new(_left: u32, _right: u32, parent: &Self::ParentType) -> Self {
```
To suppress cargo warnings about unused variables

Jakob37 commented:
Ready for another look @brentp

brentp commented:
I pushed v0.3.11 with these changes.
